### PR TITLE
Add support for LaTeX inline and block expressions.

### PR DESCRIPTION
**Pull request by luisbc92**
### Summary

This PR addresses #1038 and adds support for including LaTeX in the document using @benrbray/prosemirror-math which provides the necessary extensions to ProseMirror.

Most of the magic comes from @tommoor's implementation at outline/rich-markdown-editor#434 with some additional changes to properly integrate it with the current code base.

I was able to make use of styled-components to introduce the required styling for math blocks in accordance to the current theme, however the styles and fonts used by KaTeX forced me to introduce css-loader and file-loader, in webpack/babel.

Please review and let me know if there are any additional changes required to accept the PR!

### Screenshots
- An example showing both inline and block expressions.
image
- The same while the inline editor is active.
image
- And with the block editor.
image
- And finally, dark mode.
image
- Both are insertable through the context menu.
image

### CODE REVIEW

**Comment by tommoor** [comment]
Unfortunately this increases the application bundle size by 25%, which is unacceptable. One of the blockers in the previous PR was the inability to load the library async like we can with mermaid.js :(

**Comment by luisbc92** [comment]
Yes I did notice that.

I think that's solvable since the size increase comes from the fonts loaded by KaTeX. The two approaches I see are:

- Integrate something like https://webpack.js.org/plugins/mini-css-extract-plugin/ to split the bundle and allow asynchronous loading.
- Import the font from KaTeX asynchronously through a style tag.

Would any of these options be acceptable?

**Comment by tommoor** [comment]
Ah that does make life easier that the js increase is only 6%, as long as the fonts are only loaded when needed it's okay – however that is achieved.

**Comment by luisbc92** [comment]
That should do it!

The styles are only loaded when a MathView gets created so its not possible for the fonts to load earlier.

File: app/editor/menus/block.tsx [hunk 1]
```diff
@@ -111,6 +112,18 @@ export default function blockMenuItems(dictionary: Dictionary): MenuItem[] {
       icon: TableIcon,
       attrs: { rowsCount: 3, colsCount: 3 },
     },
+    {
+      name: "math_inline",
+      title: dictionary.mathInline,
+      icon: MathIcon,
+      keywords: "math latex",
```
Comment (tommoor) [inline app/editor/menus/block.tsx:119]:
```suggestion
      keywords: "math katex latex",
```

File: app/editor/menus/block.tsx [hunk 2]
```diff
@@ -111,6 +112,18 @@ export default function blockMenuItems(dictionary: Dictionary): MenuItem[] {
...
+    },
+    {
+      name: "math_display",
+      title: dictionary.mathDisplay,
+      icon: MathIcon,
+      keywords: "math latex block",
```
Comment (tommoor) [inline app/editor/menus/block.tsx:125]:
```suggestion
      keywords: "math katex latex",
```

File: app/editor/menus/block.tsx [hunk 3]
```diff
@@ -111,6 +112,18 @@ export default function blockMenuItems(dictionary: Dictionary): MenuItem[] {
...
       attrs: { rowsCount: 3, colsCount: 3 },
     },
+    {
+      name: "math_inline",
+      title: dictionary.mathInline,
+      icon: MathIcon,
+      keywords: "math latex",
+    },
+    {
+      name: "math_display",
```
Comment (tommoor) [inline app/editor/menus/block.tsx:122]:
```suggestion
      name: "math_block",
```

Ideally we'd name this `math_block` to match `code_block`

File: app/hooks/useDictionary.ts [hunk 4]
```diff
@@ -69,6 +69,8 @@ export default function useDictionary() {
       strong: t("Bold"),
       subheading: t("Subheading"),
       table: t("Table"),
+      mathInline: t("Math In-Line (LaTeX)"),
```
Comment (tommoor) [inline app/hooks/useDictionary.ts:72]:
```suggestion
      mathInline: t("Math Inline (LaTeX)"),
```

File: shared/editor/nodes/MathDisplay.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import { NodeSpec, NodeType, Node as ProsemirrorNode } from "prosemirror-model";
+import { EditorState } from "prosemirror-state";
+import { MarkdownSerializerState } from "../lib/markdown/serializer";
+import mathRule from "../rules/math";
+import { Dispatch } from "../types";
+import Node from "./Node";
+
+export default class MathDisplay extends Node {
+  get name() {
+    return "math_display";
```
Comment (tommoor) [inline shared/editor/nodes/MathDisplay.ts:16]:
```suggestion
    return "math_block";
```

File: shared/editor/nodes/MathDisplay.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    state.text(node.textContent, false);
+    state.ensureNewLine();
+    state.write("$$");
+    state.closeBlock(node);
+  }
+
+  parseMarkdown() {
+    return {
+      node: "math_display",
+      block: "math_display",
```
Comment (tommoor) [inline shared/editor/nodes/MathDisplay.ts:49]:
```suggestion
      node: "math_block",
      block: "math_block",
```

File: shared/editor/plugins/Math.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        activeNodeViews: value.activeNodeViews,
+        macros: value.macros,
+        prevCursorPos: oldState.selection.from,
+      };
+    },
+  },
+  props: {
+    nodeViews: {
+      math_inline: createMathView(false),
+      math_display: createMathView(true),
```
Comment (tommoor) [inline shared/editor/plugins/Math.ts:71]:
```suggestion
      math_block: createMathView(true),
```

File: shared/editor/rules/math.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    if (state.src.slice(pos, max).trim().slice(-2) === "$$") {
+      lastPos = state.src.slice(0, max).lastIndexOf("$$");
+      lastLine = state.src.slice(pos, lastPos);
+      found = true;
+    }
+  }
+
+  state.line = next + 1;
+
+  const token = state.push("math_display", "math", 0);
```
Comment (tommoor) [inline shared/editor/rules/math.ts:164]:
```suggestion
  const token = state.push("math_block", "math", 0);
```

File: shared/editor/rules/math.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    state.getLines(start + 1, next, state.tShift[start], true) +
+    (lastLine && lastLine.trim() ? lastLine : "");
+  token.map = [start, state.line];
+  token.markup = "$$";
+  return true;
+}
+
+export default function markdownMath(md: MarkdownIt) {
+  md.inline.ruler.after("escape", "math_inline", mathInline);
+  md.block.ruler.after("blockquote", "math_display", mathDisplay, {
```
Comment (tommoor) [inline shared/editor/rules/math.ts:177]:
```suggestion
  md.block.ruler.after("blockquote", "math_block", mathDisplay, {
```

File: shared/editor/components/Styles.ts [hunk 10]
```diff
@@ -11,6 +11,112 @@ export type Props = {
...
+	min-width: 1em;
+	min-height: 1em;
+	font-size: 0.95em;
+	font-family: "Consolas", "Ubuntu Mono", monospace;
+	cursor: auto;
+}
+
+.math-node.empty-math .math-render::before {
+	content: "(empty math)";
+	color: red;
```
Comment (tommoor) [inline shared/editor/components/Styles.ts:29]: Can use the red from the theme here

File: shared/editor/components/Styles.ts [hunk 11]
```diff
@@ -11,6 +11,112 @@ export type Props = {
...
+}
+
+.math-node .math-render.parse-error::before {
+	content: "(math error)";
+	color: red;
```
Comment (tommoor) [inline shared/editor/components/Styles.ts:34]: Can use the red from the theme here

File: shared/editor/components/Styles.ts [hunk 12]
```diff
@@ -11,6 +11,112 @@ export type Props = {
...
+
+math-inline {
+  display: inline; white-space: nowrap;
+
+}
+
+math-inline .math-render { 
+	display: inline-block;
+	font-size: 0.85em;
+	cursor:pointer;
```
Comment (tommoor) [inline shared/editor/components/Styles.ts:59]:
```suggestion
```

File: shared/editor/components/Styles.ts [hunk 13]
```diff
@@ -11,6 +11,112 @@ export type Props = {
...
+}
+
+math-inline .math-src .ProseMirror {
+	display: inline;
+	/* Necessary to fix FireFox bug with contenteditable, https://bugzilla.mozilla.org/show_bug.cgi?id=1252108 */
+  border-radius: 4px;
+  border: 1px solid ${props.theme.codeBorder};
+  /* background: ${props.theme.codeBackground}; */
```
Comment (tommoor) [inline shared/editor/components/Styles.ts:67]:
```suggestion
```

File: shared/editor/components/Styles.ts [hunk 14]
```diff
@@ -11,6 +11,112 @@ export type Props = {
...
+math-display .math-src .ProseMirror {
+	width: 100%;
+	display: block;
+}
+
+math-display .math-src::after, math-display .math-src::before {
+	content: "$$";
+	text-align: left;
+	color: ${props.theme.codePunctuation};
+}
```
Comment (tommoor) [inline shared/editor/components/Styles.ts:103]:
```suggestion
```

File: shared/editor/components/Styles.ts [hunk 15]
```diff
@@ -11,6 +11,112 @@ export type Props = {
...
+const mathStyle = (props: Props) => `
+/* Based on https://github.com/benrbray/prosemirror-math/blob/master/style/math.css */
+
+/* == Math Nodes ======================================== */
+
+.math-node {
+	min-width: 1em;
+	min-height: 1em;
+	font-size: 0.95em;
+	font-family: "Consolas", "Ubuntu Mono", monospace;
```
Comment (tommoor) [inline shared/editor/components/Styles.ts:23]:
```suggestion
	font-family: ${props.theme.fontFamilyMono};
```

File: shared/editor/components/Styles.ts (hunk 14 above)
Comment (tommoor) [reply shared/editor/components/Styles.ts:103]: I see the logic behind them but it doesn't follow how other blocks work in Outline to show this

File: shared/editor/components/Styles.ts (hunk 12 above)
Comment (tommoor) [reply shared/editor/components/Styles.ts:59]: A bit confusing, especially in read-only mode.

**Review by tommoor — COMMENTED** [review]
Pretty great, works with collaborative editing too 😄 

- [x] Lets icrease the padding in the math editor to match that of code blocks – ideally these should be as visually similar as possible.
- [ ] I think we should probably remove inline math from the block menu as the interaction isn't great leaving the "(empty math)" node and the block menu only works at the beginning of lines anyway?
- [ ] Need to add these new `$$` shortcuts to the list in `<KeyboardShortcuts />` component

**Random thought**, we should probably update Mermaid blocks to work the same way for consistency where they show the source when focused or read-only and the generated image otherwise.

File: shared/editor/components/Styles.ts (hunk 14 above)
Comment (luisbc92) [reply shared/editor/components/Styles.ts:103]: Agreed on consistency with the styling. However, you can also trigger a block by typing $$<space>, but I think its enough to have it on the KeyboardShortcuts component.

File: shared/editor/components/Styles.ts (hunk 11 above)
Comment (luisbc92) [reply shared/editor/components/Styles.ts]: Got it on the next commit.

File: app/editor/menus/block.tsx (hunk 1 above)
Comment (luisbc92) [reply app/editor/menus/block.tsx]: Oops, forgot to commit this one, but I'll leave it for the next commit.

**Comment by luisbc92** [comment]
You are right, it doesn't make much sense to have inline math on the menu.

I also refactored MathDisplay to MathBlock for consistency, and got the shortcuts on the component.

And yes! It'd be nice for Mermaid to show up the same way, giving you a nice clean look when not editing.

**Comment by luisbc92** [comment]
Is there anything I can do to help this feature get into mainline sooner?

**Review by tommoor — COMMENTED** [review]
Sorry for delay, gotta make sure it's just right and that takes some testing – a couple of small things here.

The other thing that's probably not a blocker, but is a bit weird – when you're editing the math block/inline the rest of the editor loses focus. It's the only node that does this, have you seen why that happens?

**Comment by luisbc92** [comment]
I hadn't noticed that. Let me look into it, I might have an idea on why that might happen.

File: shared/editor/components/Styles.ts [hunk 16]
```diff
@@ -11,6 +11,106 @@ export type Props = {
...
+
+/* -- Inline Math --------------------------------------- */
+
+math-inline {
+  display: inline; white-space: nowrap;
+
+}
+
+math-inline .math-render { 
+	display: inline-block;
```
[the remark this answers is not in the document]
Comment (luisbc92) [reply shared/editor/components/Styles.ts:57]: Sure!

**Comment by luisbc92** [comment]
I haven't been able to replicate any loss of focus, this is what I've tried:

- Insert math inline by typing $E=mc^2$, then going in and out of it with arrow keys.
- Clicking on math inline, then going in and out with arrow keys.
- Insert code block from menu (using mouse) and start typing, then going in and our with arrow keys.
- Insert code block from menu (triggered by /) and start typing, then ...
- Inserting multiple inline/blocks interspersed with other blocks and moving cursor around.

Do you have a specific scenario where you found this?

**Comment by tommoor** [comment]
We have a lot of styles that are only applied when the editor is in focus, for example the "H1" markers infront of headings (create a few headings in your doc to see), it's a clue that you're in editing mode.

When moving into a latex block these disappear as the entire editor seems to lose the appropriate classes.

**Comment by luisbc92** [comment]
Found it.

**prosemirror-math** is creating an inner **EditorView** for editing LaTeX which steals focus from the outer ProseMirror. I think the only way to solve this is by re-implementing the **NodeView** provided by **prosemirror-math** and use a different method to toggle editing, which I could do in the future.

At that point however, it might make more sense to bring that code over and remove the dependency, but I'd rather leave that for later if the drawback is minimal.

**Comment by TerenceLiu98** [comment]
@luisbc92 If I want to try the `latex-support` version, should I clone the [latex-support](https://github.com/luisbc92/outline/tree/latex-support) and build the image locally? :)

**Comment by luisbc92** [comment]
@TerenceLiu98 yes, you should be able to build it and run it locally. Be sure to follow the instructions [here](https://app.getoutline.com/s/770a97da-13e5-401e-9f8a-37949c19f97e/doc/local-development-5hEhFRXow7) to setup your environment correctly.

**Comment by TerenceLiu98** [comment]
> @TerenceLiu98 yes, you should be able to build it and run it locally. Be sure to follow the instructions [here](https://app.getoutline.com/s/770a97da-13e5-401e-9f8a-37949c19f97e/doc/local-development-5hEhFRXow7) to setup your environment correctly.

Thanks! I first only build the `Dockerfile` without building the `Dockerfile.base`, but then I figure it out.

**Comment by tommoor** [comment]
Moving this work to a branch in the main repo for final touchup.

**Comment by luisbc92** [comment]
Excellent! Thanks Tom.

**Comment by jpedrego** [comment]
Dear Outline Team. Thanks for including Latex into the Outline Markdown.

However, in *my* ideal world, I would use an offline Markdown editor to prepare the documentation and then simply copy/paste on Outline to share it with colleagues for review or use when it is ready.

Such work flow can not be currently (easily)  implemented due to two reasons:
- First, Outline decided to choose the use of $$ x $$ for inline and $$$ x $$$ for blocks, breaking away from the standard $x$ for inline and $$ x $$ for blocks that everyone else uses. I am surprise by such decision, but maybe there is a hidden technical reason...

-Second: when I have tried doing copy/paste from Markdown formatted text (after changing $$ => $$$ and $ => $$) the Latex parts are not recognized and the $ symbols are simply displayed as dollar symbols.


I am aware that I am being picky but still... 

In any case, if anyone could address the failure when copy pasting Markdown with equations, it would be awesome!

Thanks in advance.

**Comment by jpedrego** [comment]
I didn't see it coming!... I guess as I never thought about the issue with the dollar sign being used as actually being a Dollar sign !!!
PS: In my defense, an European writing here... ;-)

**Comment by tommoor** [comment]
> However, GitHub and the others clearly did not think it was a problem...

They use plain markdown, not a rich editor – so they don't have the same problems.

> The problem is that, when copy/pasting the text to Outline, the $$ and the $$$ do not seem to be taken into account....

This sounds like it should be filed as a bug

**Comment by jpedrego** [comment]
It seems like the copy/paste is working today. No idea what I did different today from the other day... In any case is working.

**Pull request merged**